src/foam/graphics/TreeNode.js

```diff
@@ -0,0 +1,297 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.graphics',
+  name: 'TreeNode',
+  extends: 'foam.graphics.Box',
+
+  requires: [
+    'foam.graphics.Box',
+    'foam.graphics.CView',
+    'foam.graphics.Label',
+    'foam.graphics.Line'
+  ],
+
+  imports: [
+    'formatNode',
+    'graph',
+    'nodeHeight',
+    'nodeWidth',
+    'padding',
+    'parentNode?',
+    'relationship'
+  ],
+  exports: ['as parentNode'],
+
+  properties: [
+    'data',
+    {
+      name: 'outline',
+      expression: function (x, nodeWidth, expanded, childNodes, padding) {
+        var nodeLeftEdgePadded = x - nodeWidth / 2 - padding / 2;
+        var nodeRightEdgePadded = x + nodeWidth / 2 + padding / 2;
+        var rootLevelOutline = [
+          {
+            left: nodeLeftEdgePadded,
+            right: nodeRightEdgePadded
+          }
+        ];
+
+        var champion = [];
+
+        for (let i = 0; i < childNodes.length && expanded; i++) {
+          // get child outline
+          // transform all rows
+          // merge levels into childoutlines
+          var childOutline = childNodes[i].outline.map(o => ({
+            left: o.left + x,
+            right: o.right + x,
+          }));
+
+          for (var j = 0; j < childOutline.length; j++) {
+            champion[j] = champion[j] || {
+              left: Number.MAX_SAFE_INTEGER,
+              right: Number.MIN_SAFE_INTEGER
+            };
+            champion[j].left = Math.min(childOutline[j].left, champion[j].left);
+            champion[j].right = Math.max(childOutline[j].right, champion[j].right);
+          }
+        }
+
+        var totalOutline = rootLevelOutline.concat(champion);
+
+        return totalOutline;
+      }
+    },
+    { name: 'height', factory: function () { return this.nodeHeight; } },
+    { name: 'width', factory: function () { return this.nodeWidth; } },
+    ['border', 'gray'],
+    {
+      name: 'childNodes',
+      factory: function () { return []; }
+    },
+    ['left', 0],
+    ['right', 0],
+    ['maxLeft', 0],
+    ['maxRight', 0],
+    {
+      name: 'x',
+      expression: function(neighborX, centerX) {
+        return neighborX + centerX;
+      },
+      postSet: function() {
+        this.neighborX = 0;
+        this.centerX = 0;
+        this.x = undefined;
+      }
+    },
+    {
+      class: 'Float',
+      name: 'neighborX'
+    },
+    {
+      class: 'Float',
+      name: 'centerX'
```

Comment:
What's the centerX for? Is it avoidable? If not, can you please add documentation explaining what this value is for?